LAM-112 Create lambda instance API + LAM-113 events for lambda instance creation

ioantsaf opened the pull request:
### LAM-112
- [x] Lambda instance creator (cluster_creator) return ansible result
- [x] Convert create and delete cluster to functions (cluster_creator)
- [x] Add authentication using token and url arguments (cluster_creator, provisioner)
- [x] Make lambda instance creation Celery task
- [x] Make lambda instance creation Django view
- [x] Write API documentation
- [x] Add PYTHONOPTIMIZE environment var to supervisord config (required for ansible to run through Celery worker)
- [x] Remove auth_url from cluster creator
- [x] Convert create instance view to POST method using Django REST framework
### LAM-113
- [x] Celery event for new lambda instance creation
- [x] Separate cluster_creator into sections (functions)
- [x] Check ansible result after each step
- [x] Call cluster_creator functions from lambda instance creation task, updating the respective status
- [x] Receive additional info from provisioner kamaki response, to put into database

ioantsaf committed:
ansible_manager return playbook run result

ioantsaf committed:
Refactor cluster_creator

ioantsaf committed:
Create lambda instance API initial commit

ioantsaf committed:
Use HTTP request headers to call cluster creation

ioantsaf committed:
Add authentication from http request to cluster provisioning

ioantsaf committed:
Fix provisioner tests

ioantsaf committed:
Return lambda instance specs json

ioantsaf committed:
Add PYTHONOPTIMIZE environment var to supervisord.conf

ioantsaf committed:
Fix minor travis issue

ioantsaf committed:
Add API documentation

ioantsaf committed:
Fix indent issue

ioantsaf commented:
Please review @gouzouni625 @themiszamani @makistsantekidis

gouzouni625 commented:
Please remove settings.py file from your pull request. If you have any changes for this file, make sure you add them to webapp/ansible/roles/service-vm/templates/settings.py.j2 since this is the file that will be installed by Ansible on setup.

ioantsaf committed:
Addition to Django settings template

ioantsaf commented:
@gouzouni625 I made the suggested changes.

themiszamani commented on webapp/api-doc/CreateLambdaInstance.md:
````diff
@@ -0,0 +1,79 @@
+The request in curl
+
+```
+curl -X GET -H "x-api-key: tJ3b3f32f23ceuqdoS_TH7m0d6yxmlWL1r2ralKcttY" -H "x-auth-url: https://accounts.okeanos.grnet.gr/identity/v2.0" -H "cloud-name: lambda" -H "master-name: test-master" -H "slaves: 1" -H "vcpus-master: 4" -H "vcpus-slave: 4" -H "ram-master: 2048" -H "ram-slave: 2048" -H "disk-master: 40" -H "disk-slave: 40" -H "ip-allocation: master" -H "network-request: 1" -H "project-name: ...
````
please write a generic example with [URL] instead of http://snf-671502.vm.okeanos.grnet.gr/backend/create_lambda_instance/

themiszamani commented on webapp/api-doc/CreateLambdaInstance.md:
```diff
@@ -0,0 +1,79 @@
+Type | Description | Required | Default value | Example value |
+------|-------------|----------|---------------|---------------|
+x-api-key | ~okeanos authentication token. If you have an account you may find the authentication token at (Dashboad-> API Access) https://accounts.okeanos.grnet.gr/ui/api_access. | `Yes` | None | tJ3b3f32f23ceuqdoS_TH7m0d6yxmlWL1r2ralKcttY
+x-auth-url | ~okeanos authentication url. If you have an account you may find the authentication url at (Dashboad-> API Access) https://accounts.okeanos.grnet.gr/ui/api_access. | `Yes` | None | https://accounts.okeanos.grnet.gr/identity/v2.0
```
why do we use the x-auth-url ?

ioantsaf replied:
It is the same everytime. Do we want it removed?

pkoro replied:
If it's the same everytime it makes more sense to omit it I think.

themiszamani commented on webapp/api-doc/CreateLambdaInstance.md:
```diff
@@ -0,0 +1,79 @@
+------|-------------|----------|---------------|---------------|
+x-api-key | ~okeanos authentication token. If you have an account you may find the authentication token at (Dashboad-> API Access) https://accounts.okeanos.grnet.gr/ui/api_access. | `Yes` | None | tJ3b3f32f23ceuqdoS_TH7m0d6yxmlWL1r2ralKcttY
+x-auth-url | ~okeanos authentication url. If you have an account you may find the authentication url at (Dashboad-> API Access) https://accounts.okeanos.grnet.gr/ui/api_access. | `Yes` | None | https://accounts.okeanos.grnet.gr/identity/v2.0
+cloud-name | The cloud name specified in the .kamakirc file | `Yes` | None | lambda
```
add these to the Parameters chapter .. 

### Parameters

| Name | Description | Required | Default value | Example value |
| --- | --- | --- | --- | --- |

ioantsaf replied:
In this API call, we do not use any URL parameters. as all variables are passed as HTTP headers. Do we want an empty 'Parameters' section?

themiszamani commented on webapp/api-doc/CreateLambdaInstance.md:
```diff
@@ -0,0 +1,79 @@
+
+# API - create lambda instance - Description
+
+Lambda instance creation call. Given an authentication token through the header x-api-key, and the instance specifications through other HTTP headers, it will firstly check the validity of the token. If the token is invalid, the API will reply with a 401 error. If the token is valid, the API will start creating a new lambda instance, using the instance specifications specified via the HTTP header ...
```
Add more details about the lambda instance creation .

ioantsaf committed:
Changes to api-doc

pkoro commented on webapp/ansible/roles/service-vm/templates/settings.py.j2:
```diff
@@ -117,9 +117,14 @@ CELERY_QUEUES = (
-            'queue': 'clusters',
-            'routing_key': 'cluster',
-        },
+    'backend.tasks.add': {       # To be removed after testing is done.
```
Does the comment apply in our case? If yes please add a TODO clause.

ioantsaf replied:
This was a test task, and will be removed.

pkoro commented on core/fokia/cluster_creator.py:
```diff
@@ -4,89 +4,109 @@
 from fokia.provisioner import Provisioner
 from fokia.ansible_manager import Manager
+# script_path = os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))
+script_path = '/var/www/okeanos-LoD/core/fokia'
```
I am not sure I like this change. I assume it is needed for testing purposes?

ioantsaf replied:
This is the path of the repo on the service VM, when it is deployed via ansible.

pkoro commented:
A couple of comments regarding this PR are the following:

1) the method is GET, but "cluster_create" makes more sense to take place under a POST verb
2) too many headers. If we can change to POST I would rather the user provides a json payload and supplies only `x-api-key` through a header.

ioantsaf commented:
Will check on implementing a POST method using json.

ioantsaf committed:
Add create_new_lambda_instance event

ioantsaf committed:
Convert create instance view to POST

Use Django REST framework
Remove auth_url from cluster creator

ioantsaf committed:
Fix settings.py.j2 indentation

ioantsaf committed:
Fix queue names, some indentation issues

ioantsaf committed:
Changed method of instance uuid generation

ioantsaf committed:
Split lambda instance creation process

Split lambda instance creation into functions. create_cluster function
creates a cluster using provisioner, while run_playbook function runs
ansible playbooks, using an ansible manager object given as argument.
Lambda instance creation celery task then sequentially calls the
functions necessary for the complete lambda instance creation.

Other tasks:
- Refactor instance creation function and library names
- Create playbook that waits for ssh, then install proxy
- Update create lambda instance api documentation

ioantsaf committed:
Added events for each ansible step

ioantsaf committed:
Added event for cluster creation step

ioantsaf committed:
Insert cluster info into database event

ioantsaf committed:
Add event to settings.py

ioantsaf committed:
Fix travis line length issue

ioantsaf committed:
Fix some dictionary parsing bugs

ioantsaf commented:
please review @makistsantekidis @pkoro

ioantsaf committed:
Fixed some celery task queue names

ioantsaf committed:
Fix celery /home/celery env var, change supervisord log paths

ioantsaf committed:
Added /home/celery env var in supervisord.conf

gouzouni625 commented:
+1
